# SNPcalling.md

The raw data is available on the High Capacity Storage of Otago University. 

Contact akstubbs.nz@gmail.com for access.

## Quality control

The data is single end __ across two lanes. 

- Plate 1 / lane1 = SQ1609...6_fastq.txt.gz
- Plate 2 / lane2 = SQ1630...7_fastq.txt.gz

The adapter content and the barcodes were subset to understand the structure before being used on whole raw data file:

```
#!/bin/sh
# in source_files_kea

ls
less Kea_conversion_table_plate1.txt

zcat SQ1609_CD82MANXX_s_6_fastq.txt.gz | head -n 1000000 > lane1_sample.fq
zcat SQ1630_CD9F4ANXX_s_7_fastq.txt.gz | head -n 1000000 > lane2_sample.fq
```

Then I check the quality of the sequeniging run using fastqc. 

```
#!/bin/sh
module load FastQC

fastqc *fq
```
Looking at the html output, sequence quality is okay but quite some error in first and last few bases.

I will first remove the last bases in each lane as there is a lot of adapter contamination. 
As we are using a reference genome for kea, we do not need to have a common read length. 
Shorter reads are removed by setting minimum number of bp to 40. 

The following uses the whole raw data file. Unzipped first using zcat.

```
#!/bin/sh
module load cutadapt

cd source_files_kea
zcat SQ1609_CD82MANXX_s_6_fastq.txt.gz > SQ1609_1.fastq
zcat SQ1630_CD9F4ANXX_s_7_fastq.txt.gz > SQ1630_2.fastq

cutadapt -j 2 -a ACCGAGATCGGAAGAGC -m 40 -o trimmed_SQ1609_1.fastq SQ1609_1.fastq 
cutadapt -j 2 -a ACCGAGATCGGAAGAGC -m 40 -o trimmed_SQ1630_2.fastq SQ1630_2.fastq 
```
```
=== Summary SQ1609 ===

Total reads processed:             261,124,595
Reads with adapters:               144,422,194 (55.3%)
Reads that were too short:          26,004,723 (10.0%)
Reads written (passing filters):   235,119,872 (90.0%)

Total basepairs processed: 26,135,975,036 bp
Total written (filtered):  18,588,542,969 bp (71.1%)

=== Summary SQ1630 ===

Total reads processed:             259,007,851
Reads with adapters:               127,761,396 (49.3%)
Reads that were too short:          19,101,143 (7.4%)
Reads written (passing filters):   239,906,708 (92.6%)

Total basepairs processed: 25,938,225,049 bp
Total written (filtered):  19,534,525,392 bp (75.3%)
```

## Reference Genome
Created a folder for kea reference genome:
```
#!/bin/sh
cd 
cd source_files_kea/
mkdir -p kea_ref_genome
```
Copied kea reference genome into kea_ref_genome folder and unzipped file

```
#!/bin/sh
curl -L -o kea_ref_genome/kea_ref_genome.fasta.gz ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/696/875/GCA_000696875.1_ASM69687v1/GCA_000696875.1_ASM69687v1_genomic.fna.gz

gunzip kea_ref_genome/kea_ref_genome.fasta.gz
```
Load BWA and index reference

```
#!/bin/sh
module load BWA

bwa index kea_ref_genome.fasta
```

## SNP calling

### Demultiplexing

First I extract barcodes from the .key file (SQ1609.txt and SQ1630.txt) of the sequencing platform.

```
#!/bin/sh
##Key provided in folder
cat SQ1609.txt | grep -E "K38" | cut -f 3-4 > barcodes_plate_1.txt
cat SQ1630.txt | grep -E "K38" | cut -f 3-4 > barcodes_plate_2.txt
```
Then I create different folders to deal with samples sequenced across the two different lanes before concatenating them together.

```
#!/bin/sh
mkdir raw1 samples1 raw2 samples2 samples_concatenated

cd ../raw1
ln -s ../source_files_kea/trimmed_SQ1609_1.fastq

cd ../raw2
ln -s ../source_files_kea/trimmed_SQ1630_2.fastq
```

I am now ready to run process_radtags to demultiplex.

Then we will have clean sample files and ready for alignment.

```
#!/bin/sh
module load Stacks

process_radtags -p raw1/ -o samples1/ -b source_files_kea/barcodes_plate_1.txt  --renz_1 pstI --renz_2 mspI -r -c -q --inline_null

process_radtags -p raw2/ -o samples2/ -b source_files_kea/barcodes_plate_2.txt  --renz_1 pstI --renz_2 mspI -r -c -q --inline_null
```

Looks good, keeping ~98% of reads **

```
Outputing details to log: 'samples1/process_radtags.raw1.log'

235119872 total sequences
  4456537 barcode not found drops (1.9%)
   394714 low quality read drops (0.2%)
   682381 RAD cutsite not found drops (0.3%)
229586240 retained reads (97.6%)

Outputing details to log: 'samples2/process_radtags.raw2.log'

239906708 total sequences
  3833078 barcode not found drops (1.6%)
   275458 low quality read drops (0.1%)
   627642 RAD cutsite not found drops (0.3%)
235170530 retained reads (98.0%)
```
Concatenate samples from the two lanes using python:

```
#python
import os
#
for sample in os.listdir ("samples2"):
  print(sample)
  if sample.endswith("gz") and not sample in os.listdir("samples1"):
    raise Exception

for sample in os.listdir ("samples2"):
  if sample.endswith("gz"):
    os.system("cat samples1/"+sample+" samples2/"+sample+" > samples_concatenated/"+sample)
```

Two samples, K38552 and K38547, were merged as they are the same bird.
```
zcat K38552.fq.gz K38547.fq.gz | gzip -c > K38552_47.fq.gz
```

### Alignment and variant calling

Make directory for bam files to be kept in:

```
#!/bin/sh
mkdir bam_files_ALLsamples
```

First alignment for every sample using BWA. Here is one example command:

```
#!/bin/sh
bwa mem -t 8 $bwa_db $src/${sample}.fq.gz | samtools view -b | samtools sort --threads 4 > ${sample}.bam
```

The complete list of commands is in [realign.sh](realign.sh):

Ran script in realign.sh

```
#!/bin/sh
cd bam_files_ALLsamples
bash ../realign.sh
```

I created a population map file for the three haplotype regions of wild kea (North, Central, South) and the captive population (unknown haplotype region).

```
#!/bin/sh
cd nano popmap_haplotype.txt
<sample file prefix><tab><population ID>

<K38*><tab><north/central/south/?>
```
I created a new folder for the output reference map:

```
#!/bin/sh
mkdir output_refmap_haplotype
```

Then I run refmap from Stacks quick run to identify low quality individuals:
Ran ref_map.pl as a job: ```sbatch ref_map.pl```

```
#!/bin/sh
ref_map.pl --samples bam_files_ALLsamples/ --popmap popmap_haplotype.txt -T 8 -o output_refmap_haplotype/
```

Now check samples and output files for low quality individuals with low sample numbers. 
No individuals with low samples numbers identified.

### Re-run ref_map populations again cleanly:

```
#!/bin/sh
module load Stacks

mkdir output_pop_haplotype

populations -P output_refmap_haplotype/ -O output_pop_haplotype/ -M popmap_haplotype.txt -p 1 --write-single-snp --plink --vcf --verbose

Removed ...
```
*Link to output files here* 
