# 2. Filtering.md

Using vcftools to remove any individuals with high missing data, and filtering SNPs.

Plink and R to create PCA plots.

## Remove individuals with high missing data

VCFtools to look at missing data per individual. Make a new directory for copying VCF file after running populations
```
#!/bin/sh
mkdir filtering_haplotype filtering_onepop
cp output_pop_haplotype/populations.snps.vcf filtering_haplotype/
cp output_pop_onepop/populations.snps.vcf filtering_onepop/
```

**vcf file in 'filtering_onepop' folder used for filtering**

Run vcftools to output a file that contains missingness on an individual level.

```
#!/bin/sh
module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1
vcftools --vcf populations.snps.vcf --missing-indv
```
Downloaded ".imiss" file and opened in Excel, sorted individuals by the most missing data (F_MISS). Created a scatterplot to visualise data (F_MISS x N_GENOTYPES-FILTERED)

Four individuals >0.70 F_MISS threshold:

K38612  K38699  K38682  K38689

  **We can remove these using the following code. However - this was not done, so we can check the filtering for all individuals. Individuals can be removed later on, after checking other filtering.**

  *Vcftools run again to output a new VCF file, which removes individuals with >0.70 missing data. New VCF file renamed appropriately. And file moved into appropriate new folder for future analysis to take place and file storage.*

```
#!/bin/sh
vcftools --vcf populations.snps.vcf --remove-indv K38612 --remove-indv K38699 --remove-indv K38682 --remove-indv K38689 --recode
mv out.recode.vcf removed_4ind_70-FMISS.vcf

mkdir filtering_70-FMISS_rm4ind
mv removed_4ind_70-FMISS.vcf filtering_70-FMISS_rm4ind/
```

  *Can also remove:  K38649 K38713 K38697 K38705* 

## Filtering

VCFtools to output the mean depth per individual:

```
#!/bin/sh
vcftools --vcf populations.snps.vcf --depth
```
Generated 'out.idepth' file. Individual with highest depth, and doubled it to obtain maximum depth parameter. ```34```

Remove SNPs with depth <2 and >34, and 80% missing data

```
#!/bin/sh
vcftools --vcf ../populations.snps.vcf --minDP 2 --maxDP 34 --max-missing 0.80 --recode
After filtering, kept 27652 out of a possible 92797 Sites
```
Rename new vcf to suitable name:

```
#!/bin/sh
mv out.recode.vcf filtered_dp2_34_md80.vcf
```
Tried for different levels of filtering - depth <2 or <3, and 70% or 80% missing data :

```
#!/bin/sh
vcftools --vcf ../populations.snps.vcf --minDP 3 --maxDP 34 --max-missing 0.80 --recode
After filtering, kept 22944 out of a possible 92797 Sites
mv out.recode.vcf filtered_dp3_34_md80.vcf

vcftools --vcf ../populations.snps.vcf --minDP 3 --maxDP 34 --max-missing 0.70 --recode
After filtering, kept 27857 out of a possible 92797 Sites
mv out.recode.vcf filtered_dp3_34_md70.vcf

vcftools --vcf ../populations.snps.vcf --minDP 2 --maxDP 34 --max-missing 0.70 --recode
After filtering, kept 32616 out of a possible 92797 Sites
mv out.recode.vcf filtered_dp2_34_md70.vcf
```
Number of SNPs kept compared. 

Stringent filtering (3-34, 0.8 = 22944), less stringent (3-34, 0.7 = 27857). Difference ~6000.

Stringent filtering (2-34, 0.8 = 27652), less stringent (2-34, 0.7 = 32616). Difference ~5000.

**Stringent filtering 3min 34max 0.8 missing ** -> still keeps 23,000 SNPs

---
# Remove sex chromosomes

Use VCFtools to remove sex chromosomes from filtered vcf file.

```
#!/bin/sh
module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1
vcftools --vcf filtered_dp3_34_md80.vcf --exclude-bed sex_chr.bed --recode
mv out.recode.vcf sex_chr_rm_filtered_dp3_34_md80.vcf
```

Full code can be found in [Sex_Chromosomes.md](https://github.com/akstubbs/keaGBS/blob/main/Sex_Chromosomes.md)

---

# PCA

## Plink

Load Plink. Give Plink the vcf (MinDP 3, missingness 0.80) and covert it into PLINK binary files (.bim,.bed,.fam). 
```
#!/bin/sh
module load PLINK/1.09b6.16

plink --vcf filtered_dp3_34_md80.vcf --allow-extra-chr --double-id --make-bed --out pca_filtered_dp3_34_md80
```
Create pca output files:
```
#!/bin/sh
plink --bfile pca_filtered_dp3_34_md80 --allow-extra-chr --pca --out pca_filtered_dp3_34_md80
```
Downloaded two output files:
- pca_filtered_dp3_34_md80.eigenvec
- pca_filtered_dp3_34_md80.eigenval

Plotting completed in R.

### Plotting PCA - detect population structure

Install appropriate R packages:

```
install.packages("tidyverse")
install.packages("readr")

library(tidyverse)
library(readr)
```

#### Base R PCA with sample names

**PC1xPC2**

```
eigenvec_table <- read.table('pca_filtered_dp3_34_md80.eigenvec')
plot(eigenvec_table[,3], eigenvec_table[,4], xlab = "PC1", ylab = "PC2", col = "seagreen", cex=0.50)
```

PCA with sample names in text for PC1xPC2
```
plot(eigenvec_table[,3], eigenvec_table[,4], xlab = "PC1", ylab = "PC2", col = "seagreen", cex=0)
text(eigenvec_table[,3], eigenvec_table[,4], eigenvec_table[,2], cex=0.50, xlim=c(-0.25,0.4))
```

<img src="https://user-images.githubusercontent.com/85653223/125231404-c818c780-e32e-11eb-82a0-6025e73fa5ff.png" width=75% height=75%>


**PC2xPC3**

```
plot(eigenvec_table[,4], eigenvec_table[,5], xlab = "PC2", ylab = "PC3", col = "seagreen", cex=0)
text(eigenvec_table[,4], eigenvec_table[,5], eigenvec_table[,2], cex=0.50, xlim=c(-0.25,0.4))
```
**PC3xPC4**

```
plot(eigenvec_table[,5], eigenvec_table[,6], xlab = "PC3", ylab = "PC4", col = "seagreen", cex=0)
text(eigenvec_table[,5], eigenvec_table[,6], eigenvec_table[,2], cex=0.35, xlim=c(-0.25,0.4))
```

### PCA with ggplot, & variances per PC

Read in data

```
pca <- read_table2("./pca_filtered_dp3_34_md80.eigenvec", col_names = FALSE)
eigenval <- scan("./pca_filtered_dp3_34_md80.eigenval")
```
Cleaning data up: remove nuisance column (had 2 col of ind names FID/ID)
```
pca <- pca[,-1]

#Set names
names(pca)[1] <- "ind"
names(pca)[2:ncol(pca)] <- paste0("PC", 1:(ncol(pca)-1))

#Convert to percentage variance explained
pve <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100)
```
Plot of PC variances
```
a <- ggplot(pve, aes(PC, pve)) + geom_bar(stat = "identity")
a + ylab("Percentage variance explained") + theme_classic()
```

<img src="https://user-images.githubusercontent.com/85653223/125231985-d3b8be00-e32f-11eb-93ee-2f9e3d6fe06b.png" width=50% height=50%>


**plot PC1 x PC2 with variance**

I wanted to be able to identify the samples by their haplotype regions (plus unknown captive) as mentioned before for the population map (ref_map and populations). 

Therefore, I extracted the pop_map file as a .csv file, and converted the V2 column (haplotypes) into factor variables that can be used to change the colour of the plot by haplotype group.

```
data = read.csv(file="haplotype-populations_ALL_individuals.csv", sep=',', header=F)
Haplotype <- as.factor(data$V2)
```

Final code for plot:

```
b <- ggplot(pca, aes(PC1, PC2, color = Haplotype)) + geom_point(size = 1)
b <- b + coord_equal() + theme_classic()
b + xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))
```

<img src="https://user-images.githubusercontent.com/85653223/125230662-5db35780-e32d-11eb-9c13-a6379009bf2f.png" width=70% height=70%>

**plot PC2 x PC3**
```
c <- ggplot(pca, aes(PC2, PC3, color = Haplotype)) + geom_point(size = 1)
c <- c + coord_equal() + theme_classic()
c + xlab(paste0("PC2 (", signif(pve$pve[2], 3), "%)")) + ylab(paste0("PC3 (", signif(pve$pve[3], 3), "%)"))
```
**plot PC3 x PC4**
```
d <- ggplot(pca, aes(PC3, PC4, color = Haplotype)) + geom_point(size = 1)
d <- d + coord_equal() + theme_classic()
d + xlab(paste0("PC3 (", signif(pve$pve[3], 3), "%)")) + ylab(paste0("PC4 (", signif(pve$pve[4], 3), "%)"))
```
___

## PCA plots excluding captive kea samples

Created file containing all captive kea sample names: kea_sample_IDs_captive.txt

Re-run Plink with flag --remove-fam to exclude samples from captive kea file.

```
plink --vcf filtered_dp3_34_md80.vcf --remove-fam ../kea_sample_IDs_captive.txt --allow-extra-chr --double-id --make-bed --out pca_filtered_dp3_34_md80_NoCaptiveInd
```

Some of the output from the plink log - showing that all the captive individuals (N=48) were successfully removed. 139 individuals remaining.
``` 
187 people (0 males, 0 females, 187 ambiguous) loaded from .fam.
Ambiguous sex IDs written to pca_filtered_dp3_34_md80_NoCaptiveInd.nosex .
--remove-fam: 139 people remaining.
```

Create PCA output files:
```
#!/bin/sh
plink --bfile pca_filtered_dp3_34_md80_NoCaptiveInd --allow-extra-chr --pca --out pca_filtered_dp3_34_md80_NoCaptiveInd
```
Downloaded two output files:
- pca_filtered_dp3_34_md80_NoCaptiveInd.eigenvec
- pca_filtered_dp3_34_md80_NoCaptiveInd.eigenval

Plotting completed in R.

**PC1xPC2**

```
eigenvec_table <- read.table('pca_filtered_dp3_34_md80_NoCaptiveInd.eigenvec')
plot(eigenvec_table[,3], eigenvec_table[,4], xlab = "PC1", ylab = "PC2", col = "seagreen", cex=0.50)
```
PCA with sample names in text for PC1xPC2
```
plot(eigenvec_table[,3], eigenvec_table[,4], xlab = "PC1", ylab = "PC2", col = "seagreen", cex=0)
text(eigenvec_table[,3], eigenvec_table[,4], eigenvec_table[,2], cex=0.50, xlim=c(-0.25,0.4))
```

**PCA with ggplot, & variances per PC**

Read in data

```
pca <- read_table2("./pca_filtered_dp3_34_md80_NoCaptiveInd.eigenvec", col_names = FALSE)
eigenval <- scan("./pca_filtered_dp3_34_md80_NoCaptiveInd.eigenval")
```
Cleaning data up: remove nuisance column (had 2 col of ind names FID/ID)
```
pca <- pca[,-1]

#Set names
names(pca)[1] <- "ind"
names(pca)[2:ncol(pca)] <- paste0("PC", 1:(ncol(pca)-1))

#Convert to percentage variance explained
pve <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100)
```

**plot PC1 x PC2 with variance of wild kea only**

I wanted to be able to identify the wild kea samples by their haplotype regions (Northern, Central, Southern). 

Therefore, I extracted the pop_map file as a .csv file, and converted the V2 column (haplotypes) into factor variables that can be used to change the colour of the plot by haplotype group.

```
data = read.csv(file="haplotype-populations_NO_captive_individuals.csv", sep=',', header=F)
Haplotype <- as.factor(data$V2)
```

Final code for wild kea **haplotype** plot:
```
b <- ggplot(pca, aes(PC1, PC2, color = Haplotype)) + geom_point(size = 1)
b <- b + coord_equal() + theme_classic()
b + xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))
```

Final code for wild kea **region** plot:
```
data = read.csv(file="regions-populations_ALL_individuals.csv", sep=',', header=F)
Region <- as.factor(data$V2)

b <- ggplot(pca, aes(PC1, PC2, color = Region)) + geom_point(size = 1)
b <- b + coord_equal() + theme_classic()
b + xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))
```
___

## PCA plots of wild kea + individuals in captive population BORN IN WILD

Captive kea born in wild: 
- K38637
- K38648
- K38664
- K38670
- K38704
- K38709

Created file containing all **captive born** kea sample names: kea_sample_IDs_captiveBORN.txt

Re-run Plink with flag --remove-fam to exclude samples from captive born kea file.

```
plink --vcf filtered_dp3_34_md80.vcf --remove-fam ../kea_sample_IDs_captiveBORN.txt --allow-extra-chr --double-id --make-bed --out pca_filtered_dp3_34_md80_NoCaptiveBORN
```

Some of the output from the plink log - showing that all the captive born individuals (N=42) were successfully removed. 139 individuals remaining.
``` 
187 people (0 males, 0 females, 187 ambiguous) loaded from .fam.
Ambiguous sex IDs written to pca_filtered_dp3_34_md80_NoCaptiveBORN.nosex .
--remove-fam: 145 people remaining.
```

Create PCA output files:
```
#!/bin/sh
plink --bfile pca_filtered_dp3_34_md80_NoCaptiveBORN --allow-extra-chr --pca --out pca_filtered_dp3_34_md80_NoCaptiveBORN
```
Downloaded two output files:
- pca_filtered_dp3_34_md80_NoCaptiveBORN.eigenvec
- pca_filtered_dp3_34_md80_NoCaptiveBORN.eigenval

Plotting completed in R.

**PC1xPC2**

```
eigenvec_table <- read.table('pca_filtered_dp3_34_md80_NoCaptiveBORN.eigenvec')
plot(eigenvec_table[,3], eigenvec_table[,4], xlab = "PC1", ylab = "PC2", col = "seagreen", cex=0.50)
```
PCA with sample names in text for PC1xPC2
```
plot(eigenvec_table[,3], eigenvec_table[,4], xlab = "PC1", ylab = "PC2", col = "seagreen", cex=0)
text(eigenvec_table[,3], eigenvec_table[,4], eigenvec_table[,2], cex=0.50, xlim=c(-0.25,0.4))
```

**PCA with ggplot, & variances per PC**

Read in data

```
pca <- read_table2("./pca_filtered_dp3_34_md80_NoCaptiveBORN.eigenvec", col_names = FALSE)
eigenval <- scan("./pca_filtered_dp3_34_md80_NoCaptiveBORN.eigenval")
```
Cleaning data up: remove nuisance column (had 2 col of ind names FID/ID)
```
pca <- pca[,-1]

#Set names
names(pca)[1] <- "ind"
names(pca)[2:ncol(pca)] <- paste0("PC", 1:(ncol(pca)-1))

#Convert to percentage variance explained
pve <- data.frame(PC = 1:20, pve = eigenval/sum(eigenval)*100)
```

**plot PC1 x PC2 with variance of wild born kea only**

I wanted to be able to identify the wild kea samples by their haplotype regions (Northern, Central, Southern), and see where the wild-born captive kea sit. 

Therefore, I extracted the pop_map file as a .csv file, and converted the V2 column (haplotypes) into factor variables that can be used to change the colour of the plot by haplotype group.


Final code for wild born kea **haplotype** plot:

```
data = read.csv(file="haplotype-populations_NO_captiveBORN_individuals.csv", sep=',', header=F)
Haplotype <- as.factor(data$V2)

b <- ggplot(pca, aes(PC1, PC2, color = Haplotype)) + geom_point(size = 1)
b <- b + coord_equal() + theme_classic()
b + xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))
```

Final code for wild born kea **region** plot:
```
data = read.csv(file="regions-populations_NO_captiveBORN_individuals.csv", sep=',', header=F)
Region <- as.factor(data$V2)

b <- ggplot(pca, aes(PC1, PC2, color = Region)) + geom_point(size = 1)
b <- b + coord_equal() + theme_classic()
b + xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))
```
