# Inbreeding estimates

## Inbreeding coefficient: FH

**PLINK v1.9 --het**

Calculate inbreeding coefficients based on the observed versus expected number of homozygous genotypes.

```
## Make bed file for filtered vcf file with sex chromosomes removed
plink --vcf sex_chr_removed.vcf --allow-extra-chr --double-id --make-bed --out ibd_test

plink --bfile ibd_test --recode --tab --allow-extra-chr --out ibd_test

plink --bfile ibd_test --allow-extra-chr --het --ibc --out ibd_test
```

**vcftools --het**

Calculate inbreeding coefficient on per-individual basis, using method of moments.
```
vcftools --vcf sex_chr_removed.vcf --het
```

Outputs from VCFtools --het and PLINK --het are the same.


## IBD estimation

**PLINK v1.9 --genome**

Useful for detecting pairs of individuals who look more similar to eachother (more than we would expect by chane in a random sample).

```
plink --bfile ibd_test --allow-extra-chr --genome --out ibd_test
```

## ROH

**PLINK v1.9 --homozyg**

```
plink --bfile ibd_test --allow-extra-chr --homozyg --homozyg-window-snp 50 --homozyg-window-missing 5 --homozyg-snp 100 --homozyg-density 50
```
Ran --homozyg with default parameters, with no ROH found. Altered parameters and still no ROH. Checked kea genome on NCBI - Scaffold N50 at 61K (sequence length of the shortest scaffold at 50% of the total genome length). Therefore, this is likely why no ROH appear (?).

Perhaps look at ROH with denovo sequencing instead?
