# Captive kea ROH analysis

We mapped the captive individuals (N=48) to the kakapo genome instead of kea genome to try ROH analysis - as this did not work when mapped to kea genome which has very low N50 scaffold length (~61Kb).

## Indexing kakapo genome 

Kakapo genome indexed:

```
#!/bin/sh
cd uoo03341/source_files_kea/kakapo_sex_chr
bwa index kakapo_ref_genome_GCA_004027225.2_bStrHab1.2.pri_genomic.fna
```

Align CAPTIVE kea to kakapo using commands in [realign_kea_to_kakapo.sh]():

```
#!/bin/sh
cd uoo03341/roh_mapped_to_kakapo_genome
bash realign_kea_to_kakapo.sh
```

Captive kea mapped to kakapo genome .bam files in uoo03341/roh_mapped_to_kakapo_genome/bam_captive_kea_to_kakapo

## Variant Calling

Run refmap from Stacks quick to identify low quality individuals

```
#!/bin/sh
mkdir output_map
ref_map.pl --samples bam_captive/ --popmap captive_popmap.txt -T 8 -o output_map/
```

Look at effective_coverages_per_sample (mean_cov) in gstacks.log.distribs file for low quality individuals (K38673  K38714  K38657  K38594  K38666)

**Rerun populations again cleanly**

```
#!/bin/sh
module load Stacks
populations -P output_map/ -O output_map/ -M captive_popmap.txt -p 1 --write-single-snp --plink --vcf --verbose
```

## Filtering

### Remove individuals with high missing data

Copy vcf file from populations into new filtering folder
```
#!/bin/sh
mkdir filtering
cp output_map/populations.snps.vcf filtering/
```

Run vcftools to output a file that contains missingness on an individual level

```
#!/bin/sh
module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1
vcftools --vcf populations.snps.vcf --missing-indv
```
Look at missingness of individuals in .imiss file (F_MISS).

Individuals with high missing data: 

Remove these individuals and output new vcf file, named appropriately.

```
#!/bin/sh
vcftools --vcf populations.snps.vcf --remove-indv --recode
mv out.recode.vcf removed_ind.vcf
```
### Filtering 

VCFtools to output the mean depth per individual:

```
#!/bin/sh
vcftools --vcf removed_ind.vcf --depth
```

Generated 'out.idepth' file. Individual with highest depth, and doubled it to obtain maximum depth parameter. ``` ```

Remove SNPs with depth <2 and > , and 80% missing data. Rename file appropriately.

```
#!/bin/sh
vcftools --vcf ../removed_ind.vcf --minDP 2 --maxDP 34 --max-missing 0.80 --recode
After filtering, kept  out of a possible  Sites
mv out.recode.vcf removed_ind_filtered_dp2_34_md80.vcf
```

Tried for different levels of filtering to compare number of SNPs kept - depth <2 or <3, and 70% or 80% missing data :

```
#!/bin/sh
vcftools --vcf ../removed_ind.vcf --minDP 3 --maxDP 34 --max-missing 0.80 --recode
After filtering, kept  out of a possible  Sites
mv out.recode.vcf removed_ind_filtered_dp3_34_md80.vcf

vcftools --vcf ../removed_ind.vcf --minDP 3 --maxDP 34 --max-missing 0.70 --recode
After filtering, kept  out of a possible  Sites
mv out.recode.vcf removed_ind_filtered_dp3_34_md70.vcf

vcftools --vcf ../removed_ind.vcf --minDP 2 --maxDP 34 --max-missing 0.70 --recode
After filtering, kept  out of a possible  Sites
mv out.recode.vcf removed_ind_filtered_dp2_34_md70.vcf
```

Chose filtering -> still keeps  SNPs

### Remove sex chromosomes

Use VCFtools to remove sex chromosomes from filtered vcf file.

```
#!/bin/sh
module load VCFtools/0.1.15-GCC-9.2.0-Perl-5.30.1
vcftools --vcf removed_ind_filtered_dp3_34_md80.vcf --exclude-bed sex_chr.bed --recode
mv out.recode.vcf removed_ind_filtered_dp3_34_md80_sex_chr.vcf
```
Output below. Left with  possible sites.
```
VCFtools
```
## Inbreeding estimates

### Inbreeding coefficient: FH

**vcftools --het**

Calculate inbreeding coefficient on per-individual basis, using method of moments.

```
#!/bin/sh
vcftools --vcf removed_ind_filtered_dp3_34_md80_sex_chr.vcf --het
```

**PLINK v1.9 --het**

Calculate inbreeding coefficients based on the observed versus expected number of homozygous genotypes.

*Make bed file for filtered vcf file with sex chromosomes removed*

```
#!/bin/sh
plink --vcf removed_ind_filtered_dp3_34_md80_sex_chr.vcf --allow-extra-chr --double-id --make-bed --out captive_ibd

plink --bfile captive_ibd --recode --tab --allow-extra-chr --out captive_ibd

plink --bfile captive_ibd --allow-extra-chr --het --ibc --out captive_ibd
```
### IBD estimation 

**PLINK v1.9 --genome**

Useful for detecting pairs of individuals who look more similar to eachother (more than we would expect by chane in a random sample).

```
#!/bin/sh
plink --bfile captive_ibd --allow-extra-chr --genome --out captive_ibd
```

### ROH

**PLINK v1.9 --homozyg**

```
#!/bin/sh
plink --bfile captive_ibd --allow-extra-chr --homozyg --homozyg-window-snp 50 --homozyg-window-missing 5 --homozyg-snp 100 --homozyg-density 50
```






