# TrioML using coancestry

See the manual for R package [related](https://github.com/timothyfrasier/related/blob/master/Tutorial.pdf).

COANCESTRY program (Wang 2011) in related R package used to estimate pairwise relatedness based on SNPs.

### Correct format of data

Data format of the genotype file is simple one column of individual identifiers, followed by genotype data - for which there are two columns for each locus (one per allele).

- Text file
- Space- or tab-delimited
- Missing data represented by 0 or NA
- No row of column names

**Also note that the file needs the alleles to be in 1/2/3/4 format not A/C/G/T**

The .ped file created by plink can be used, but will need to be recoded with ```--allele1234``` instead:

```
plink --bfile ../final_filtered_43captive --allow-extra-chr --allele1234 --recode coancestry_rm5cap
```

Final genotype input should file looks like the below, however columns 2-6 need to be removed from .ped file.

```
K38540  1 2 3 4 1 2 3 4 ...
K38541  2 3 1 4 1 4 2 3 ...
K38542  1 3 4 4 1 4 2 3 ...
...
```

### Reading in data to R

Remove columns 2-6 which contain
```
library(related)
genotable <- read.table("coancestry_rm5cap.ped", header=F, stringsAsFactors=F)
head(genotable[1:5,1:10])
genotable$V2 <- NULL
genotable$V3 <- NULL
genotable$V4 <- NULL
genotable$V5 <- NULL
genotable$V6 <- NULL
head(genotable[1:5,1:10])
```
