# Hierfstat for genetic diversity 

Followed instructions in [importing data in hierfstat](https://www2.unil.ch/popgen/teaching/WISG17/import.pdf) document for VCF files.

Use .vcf file and population group .txt file 

Install and load packages in R:

```
R

library("gaston")
library("pegas")
library("hierfstat")
```

Load sample descriptions (populations)

```
ind.desc <- read.table("popmap_haplo_nums183.txt",header=F)
head(ind.desc)
table(ind.desc$V2)
```

Load .vcf file:
```
fn <- "rm4highmiss_final.vcf"
test <- VCFloci(fn)

base<-c("A","T","G","C")
snps<-which(test$REF %in% base & test$ALT %in% base)

z <- read.vcf(fn, which.loci = snps)
```

**z** is a object of class loci from the *pegas* package. It can be transformed in a *hierfstat* data
frame using first ```loic2genind``` and then ```genind2hierfstat```. Since there is no pop slot in the loci
object we have just created, we need to specify explicitly the pop argument to ```genind2hierfstat```:

```
abc <- genind2hierfstat(loci2genind(z), pop = ind.desc$V2) 

all.equal(rownames(abc), as.character(ind.desc$V1))
```

# diveRsity for genetic diversity

Output genepop file from Stacks populations
```
cd uoo03341/test_rerun_filtering/all_ind/rm4highmiss_all_ind/
module load Stacks
populations -V rm4highmiss_final.vcf.gz -O . --write-single-snp --genepop --popmap ../../../popmap_haplotype.txt
```

```
cd ~/uoo03341/test_rerun_filtering/all_ind/rm4highmiss_all_ind/genepop 
module purge
module load R/4.0.1-gimkl-2020a
R
```
```
install.packages("diveRsity")
library(diveRsity)
divBasic(infile = "~/uoo03341/test_rerun_filtering/all_ind/rm4highmiss_all_ind/genepop/rm4highmiss_final.p.snps.genepop", outfile = "diversity_output", bootstraps=10)
```



