# Unique alleles

R script to run unique alleles created by Dr. Ludovic Dutoit, University of Otago, 28/10/2021.

Plotting script created by Aimee Stubbs, University of Otago, 15/11/2021.

R script can be found at ...

## Running uniquevariation.R

- vcf file
- reference/background list with one individual sample name per line. 
- an output folder needs to exist 

***The background list can be changed depending on question being asked.***

For example, changing from a full list of samples to a list containing a subset from one population - such as a:

**Captive only subset**
- Means we can find which individuals in wild population have more unique diveristy compared to the captive population (which can be used to identify locations with more 
'uniqueness' that is not in captive population - *i.e. unique diversity that is not represented*). 
- Number of unique SNPs for captive individuals when using this list is expected to be lower than wild population because of higher change of allele sharing between more 
related individuals in the captive population (i.e. majority are full-sibling relationships).

**Wild only subset**
- Means we can identify which wild individuals or areas in the wild population have more unique diversity relative to the rest of the wild population (i.e. doesn't include 
captive population representation). 
- Can also identify which captive individuals are more 'unique' compared to the wild population - therefore accounting for the likely high relatedness of the captive population 
and thus increased allele sharing betweeen closer related individuals (compared to wild population by chance).

## Plotting results

Want to plot number of unique SNPs per indiviual compared to genetic diversity (observed heterozygosity), inbreeding coefficient, missingness, AND ***latitude***

For example below plot shows wild individuals with more unique SNPs compared to the captive population tend to have a lower latitude (i.e. southern wild inividuals have more 
unique SNPs than northern kea, and therefore are less represented in captive population).  

Relationship between number of unique SNPs per wild individual (N=135) and sampling latitude (when background list is of captive birds). Colour corresponds to sampling latitude.

![Captive_num-uni_lat](https://user-images.githubusercontent.com/85653223/146846163-1a183efa-5fae-43b5-a878-2e9025d4b543.png)

________________

## Testing random list of 30 individuals *per cluster* (assigned by structure Q values) for unique alleles

Wanting to check that # individuals per cluster does not effect # of unique alleles chosen per cluster (using the clusters that were assigned by structure Q-value >0.75 per cluster).

For example - we have 68 wild indivduals assigned to northern cluster, 36 to southern, and 31 to admixed (- plus 48 in captive)... 
so we want to check that having double the number of individuals from northern cluster does not change the low number of unique alleles found in the northern population. 

They just have lower genetic diversity which is why we see less unique alleles - as opposed to the sample size (i.e. theoretically the more samples you have from one region the less
unique they will be as they are closer together anyway so are more likely to be related to some extent - so we want to remove this variable)...

### Create list of random 30 individuals for each cluster:

Using separate lists of individuals for each cluster to randomly choose 30 individals from list, before concatenating separate lists into one to be kept in .vcf file.

```
cd uoo03341/unique_alleles_code/

shuf -n 30 north-structure-cluster.txt > northcluster_rand30.txt
shuf -n 30 south-structure-cluster.txt > southcluster_rand30.txt
shuf -n 30 admixed-structure-cluster.txt > admixcluster_rand30.txt
shuf -n 30 captive-structure-cluster.txt > captivecluster_rand30.txt

cat *30.txt > rand30percluster1.txt

module load VCFtools
vcftools --vcf ../test_rerun_filtering/all_ind/rm4highmiss_all_ind/rm4highmiss_final.vcf --keep rand30percluster1.txt --recode
```
Prints below output:
```
Parameters as interpreted:
        --vcf ../test_rerun_filtering/all_ind/rm4highmiss_all_ind/rm4highmiss_final.vcf
        --keep rand30percluster1.txt
        --recode

Keeping individuals in 'keep' list
After filtering, kept 120 out of 183 Individuals
Outputting VCF file...
After filtering, kept 23049 out of a possible 23049 Sites
Run Time = 47.00 seconds
```
Rename
```
mv out.recode.vcf random30ind_perclust1.vcf
```

***Can repeat this up to 10 times and check variance of the different vcf lists.***

### Rerun uniquevariation.R

Now use the new .vcf file in the unique alleles R code, and use the 'rand30percluster1.txt' file as the background list? 

**OR** potentially use full captive list as a background instead (?) - test if this works... by using captive background list find if wild individuals/areas not represented
in the wild. To be mapped longitudinally.

Use wild background list to see which individuals in wild have greater genetic diversity - is it still the southern kea?
