# Pedigree reconstructuion using sequoia

See review [Flanagan & Jones (2018)](https://onlinelibrary.wiley.com/doi/10.1111/mec.14988) on parentage analysis.

>*"If many relatives are likely to be present in the sample, one solution might be to use an approach that estimates a broader pedigree than just parentâ€“offspring relationships. 
Programs such as **sequoia** (Huisman, 2017)... can be used to perform these sorts of analyses."*

Using sequoia package in R for pedigree reconstruction ([Huisman 2017](https://onlinelibrary.wiley.com/doi/full/10.1111/1755-0998.12665#citedby-section)).

```
install.packages('sequoia')
library('sequoia')
```

## Input files

### Genotypes
Use PLINK to create genotype data that can be used by sequoia:

*"Parentage assignment with SNP data is optimised with fewer SNPs in low linkage disequilibrium, with low missing data and a high MAF, 
so SNPs are more informative and more computationally efficient (Huisman 2017)"* (Farquharson et al. 2019)
-> ```--geno 0.2``` and ```--maf 0.01```/```0.1```

Filtered SNPs with a missingness below 10%, a MAF of at least 30%, and which in a window of 50 SNPs, sliding by 5 SNPs per step,
have a VIF (variance inflation vactor) of maximum 2. 
See [PLINK documentation](https://zzz.bwh.harvard.edu/plink/summary.shtml#prune) for choice of parameters.

```
module load PLINK/1.09b6.16

plink --bfile final_filtered_43captive --allow-extra-chr --geno 0.1 --maf 0.3 --indep 50 5 2
plink --bfile final_filtered_43captive --allow-extra-chr --extract plink.prune.in --recodeA --out inputfile_for_sequoia_43cap
```
The ```plink.prune.in``` list can be used to create the genotype file. 

This creates a file with extention .raw that containted **2723 SNPs** which can be converted to required input format in sequoia (detailed below)

### Life History Data

Created a .csv file with life history data for all captive individuals.

Format of .csv file, three - five columns:
-   **ID**: individual ID
-   **Sex**: 1=Female, 2=Male
-   **BirthYear**: Birth year if known   
-   **BY.min** (optional): If birth year unknown, earliest year individual may have been born
-   **BY.max** (optional): If birth year unknown, latest year individual may have been born

## Running sequoia

```
R

library('sequoia')

GenoM <- GenoConvert(InFile = "inputfile_for_sequoia_43cap.raw", InFormat="raw")
LifeHistData <- as.data.frame(read.csv("sequoia_info.csv", header = T))
```

Check input is in correct format:
```
CheckGeno(GenoM)
```

#### Pedigree reconstruction:
```
SeqOUTfull <- sequoia(GenoM = GenoM, 
                  LifeHistData = LifeHistData, 
                  Err = 0.005,
                  Module="ped", 
                  quiet="verbose", 
                  Plot=TRUE)
```
- Also increased error to 0.01 (as in Farquharson et al. 2019)

#### Compare sequoia pedigree to studbook pedigree
```
compare <- PedCompare(
  Ped1 = Pedigree,
  Ped2 = SeqOUT$Pedigree,
  DumPrefix = c("F0", "M0"),
  SNPd = NULL,
  Symmetrical = TRUE,
  minSibSize = "2sib",
  Plot = TRUE
  )
```

